tracer/src/Datadog.Trace/Debugger/Symbols/SymbolsUploader.cs

```diff
@@ -0,0 +1,289 @@
+        {
+            _isSymbolUploaderEnabled = null;
+            _alreadyProcessed = new HashSet<string>();
+            _environment = tracerSettings.EnvironmentInternal;
+            _serviceVersion = tracerSettings.ServiceVersionInternal;
+            _serviceName = serviceName;
+            _discoveryService = discoveryService;
+            _api = api;
+            _assemblySemaphore = new SemaphoreSlim(1);
+            _discoveryServiceSemaphore = new SemaphoreSlim(0);
+            _thresholdInBytes = settings.SymbolBatchSizeInBytes;
+            _cancellationToken = new CancellationTokenSource();
+            _jsonSerializerSettings = new JsonSerializerSettings { NullValueHandling = NullValueHandling.Ignore };
+            _discoveryService.SubscribeToChanges(ConfigurationChanged);
+        }
+
+        private void ConfigurationChanged(AgentConfiguration configuration)
+        {
+            _isSymbolUploaderEnabled = configuration.SymbolDbEndpoint;
+            if (string.IsNullOrEmpty(Volatile.Read(ref _isSymbolUploaderEnabled)))
+            {
+                Log.Information("Dynamic Instrumentation failed to upload symbols for auto-complete suggestions. This can happen if your datadog-agent version is less than 7.45");
+                return;
+            }
+
+            _discoveryServiceSemaphore.Release(1);
+            _discoveryService!.RemoveSubscription(ConfigurationChanged);
+            _discoveryService = null;
+        }
+
+        public static ISymbolsUploader Create(IBatchUploadApi api, IDiscoveryService discoveryService, DebuggerSettings settings, ImmutableTracerSettings tracerSettings, string serviceName)
+        {
+            if (api is not NoOpSymbolBatchUploadApi &&
+               (EnvironmentHelpers.GetEnvironmentVariable(ConfigurationKeys.Debugger.SymbolDatabaseUploadEnabledInternal, "false")?.ToBoolean() ?? false))
+            {
+                return new SymbolsUploader(api, discoveryService, settings, tracerSettings, serviceName);
+            }
+
+            Log.Information("Symbol database uploading is disabled. To enable it, please set {EnvironmentVariable} environment variable to 'true'.", ConfigurationKeys.Debugger.SymbolDatabaseUploadEnabled);
+            return new NoOpUploader();
+        }
+
+        private void RegisterToAssemblyLoadEvent()
+        {
+            AppDomain.CurrentDomain.AssemblyLoad += async (_, args) =>
+            {
+                await ProcessItemAsync(args.LoadedAssembly).ConfigureAwait(false);
+            };
+        }
+
+        private async Task ProcessItemAsync(Assembly assembly)
+        {
+            await Task.Yield();
+            await _assemblySemaphore.WaitAsync(_cancellationToken.Token).ConfigureAwait(false);
+
+            if (_cancellationToken.IsCancellationRequested)
+            {
+                return;
+            }
```

Comment:
For completeness: why the `_assemblySemaphore` is not getting released?

Reply:
Fixed in f74ef445ed4a6220e0849d0be2ba5866e48a3fa5